### ISSUE TITLE
Support for constraints in chronos

### ISSUE THREAD

**Issue by mattomatic** [mesos/chronos#256]
Is it possible to add support to chronos for constraints, such as those that marathon already supports (https://mesosphere.github.io/marathon/docs/constraints.html)?

For example, it would be very useful to be able to schedule jobs to be run in a particular data center. 

Also, is there an existing workaround for this?

**Comment by elingg** [comment]
Adding constraints would be an important feature for Chronos. Patches from the community would be excellent if we aren't able to add this in our immediate roadmap.

**Comment by veverjak** [comment]
Having marathon-like constrains feature would be super cool - https://mesosphere.github.io/marathon/docs/constraints.html

**Comment by kutchar** [comment]
+1

**Comment by mjbroekman** [comment]
+1 ... having the ability to specify parts of your cluster to run jobs on would great...

**Comment by synergyhue** [comment]
+1

**Comment by scaldmaster** [comment]
+1

**Comment by siddharth178** [comment]
+1

**Comment by freshmatrix** [comment]
+1

**Comment by SEJeff** [comment]
Hey gents, instead of a pointless +1, just click the Subscribe button on the top right. More and more +1s are not very useful and just spam the developers unnecessarily.

**Comment by ssk2** [comment]
@SEJeff, the +1s can be helpful for prioritisation. Ideally we'd have a less spammy system or the ability to vote built into GH issues, of course.

**Comment by SEJeff** [comment]
That is what subscribe is for :)

I comaintain a few largish open source projects (> 500 unique contributors each), salt and graphite. It works pretty well.

**Comment by gyllen** [comment]
@brndnmtthws How hard would this be to accomplish, do you have any idea. It is most certainly an important feature.

**Comment by kapil-malik** [comment]
This works pretty well now for us in latest version. Curious why this issue is still open now?

**Comment by potto007** [comment]
Isn't it still restricted to just the EQUALS constraint at this point? I
think the intent is to bring it to parity with Marathon wrt constraints.

On Thu, Jun 18, 2015, 7:53 PM KAPIL MALIK <notifications@github.com> wrote:

> This works pretty well now for us in latest version. Curious why this
> issue is still open now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesos/chronos/issues/256#issuecomment-113338593>.
>
-- 

Paul Otto
Principal Software Architect
Otto Ops LLC | *OttoOps.com*
970.343.4561 office
720.381.2383 cell

**Comment by kapil-malik** [comment]
Oh yes, agreed.EQUALS was serving our purpose, didn't realize the full story. Thanks.

**Comment by goodwillcoding** [comment]
Is there a way to set constraints using the gui?

**Comment by ssk2** [comment]
@goodwillcoding - not yet but we're working on revamping the GUI!

**Comment by tony-kerz** [comment]
is there any support for running a job on __every__ slave in a cluster?

**Comment by brndnmtthws** [comment]
Not yet.

**Comment by tony-kerz** [comment]
is something like below supposed to work?:

```
  "constraints": [
    ["hostname", "EQUALS", "ip-10-0-0-95.ec2.internal"]
  ],
```

**Comment by solarkennedy** [comment]
Yes. `EQUALS` is one of the supported constraints per the docs: https://mesos.github.io/chronos/docs/api.html#constraints

**Comment by tony-kerz** [comment]
thx @solarkennedy, i saw that, but it wasn't clear to me if `hostname` was an attribute that was auto-magically made available to chronos for use with `EQUAL`/`LIKE` constraints or if those would only work on attributes that i manually punched into a slave via something like `--attributes` (which i found somewhere in the mesos doc), or `MESOS_ATTRIBUTES` (which i see in the cloud-formation file provided by mesosphere).

looks like marathon is pretty good about [setting up `hostname` for use with constraints](https://mesosphere.github.io/marathon/docs/constraints.html), but i saw no such info in the chronos doc.

**Comment by solarkennedy** [comment]
That is true, the docs do assume that the attribute used by the constraint is declared. As far as I know hostname is not a magic attribute that you get for free.

Can you make a PR to clarify that and cross-link to the mesos docs on how to add attributes? If not I could do it.

**Comment by tony-kerz** [comment]
@solarkennedy i can take a whack at clarifying the docs. 

more importantly tho, i think it would be great if `hostname` was made available by default for constraints as it is in marathon. do you think this would be a valuable feature? if so, do you think it is (or should be) implied by this particular issue, or do you think it should be spiked out as a sep enhancement request?

**Comment by solarkennedy** [comment]
I'm not sure. I was not aware that that was a "free" constraint that you could use from marathon, even if you didn't set that as a mesos attribute. (I assumed the constraint language only operated on mesos attributes)

**Comment by tony-kerz** [comment]
check the last sentence of the first paragraph at this link: https://mesosphere.github.io/marathon/docs/constraints.html

> The field can be the slave hostname or any Mesos slave attribute.

i've constrained to `hostname` successfully in marathon as well without doing anything special.

**Comment by solarkennedy** [comment]
Ah! Today I learned. @tony-kerz for this specific feature I recommend making a separate issue/pr. 

If you ask me this main issue should be closed, as there is *some* constraint functionality. Any new feature requests or doc updates and be new things.

**Comment by anhcuong** [comment]
Please support the hostname constraint in the future because I guess many people are using Marathon along with Chronos on top of Mesos. And Marathon already supported hostname constraint

**Comment by tony-kerz** [comment]
fyi @anhcuong i have had #563 open for a while to provide that feature.

**Comment by alex-hoang-piksel** [comment]
Is there any news on adding the constraints field to the UI?
Currently when I set a constraint on a job, if the job is then edited and saved via the UI, the constraints field is emptied.

**Comment by anhcuong** [comment]
@alex-hoang-piksel you may want to check out Airflow (https://airflow.incubator.apache.org/). They provided many more features with a rich UI. And yeah, the repository is more active than Chronos.
